### PR TITLE
Olmo3 checkpoint conversion and Refactor Olmo3 model to support interleaved RoPE (and attention)

### PR DESCRIPTION
**Pull request by gagika**
# Description

This PR adds full support for AllenAI's **Olmo3-7B** and **Olmo3-32B** models in MaxText, including checkpoint conversion and forward pass correctness verification.

This implementation addresses several unique architectural features of Olmo3 that required changes to the core layers:

1.  **Global QK Normalization**: Olmo3 applies RMSNorm across the entire hidden dimension (e.g., 4096) before splitting into heads, whereas MaxText's default `RMSNorm` applies per-head.
    * *Change*: Updated `src/MaxText/layers/attentions.py` to reshape query/key tensors `[B, L, H, D] -> [B, L, H*D]` before normalization when `is_olmo3` is detected.

2.  **Mixed RoPE Strategy**: Olmo3 uses a hybrid positional embedding strategy where "Sliding Window" layers use standard RoPE, while "Global" layers use YaRN.
    * *Change*: Updated `src/MaxText/layers/olmo3.py` to explicitly override the `rope_type` to `"default"` for local sliding layers.
    * *Change*: Updated `src/MaxText/layers/attentions.py` to accept a `rope_type` override in `__init__`, enabling layer-specific RoPE configurations.

3.  **Configuration Alignments**:
    * Set `rope_interleave: False` to match Hugging Face's concatenated RoPE.
    * Set `rope_truncate: False` to prevent frequency drift in YaRN layers.
    * Set `normalize_embedding_logits: False` as Olmo3 does not normalize output logits.
    * Renamed config files to use hyphens (`olmo3-7b.yml`) to match standard naming conventions.

4.  **Checkpoint Conversion**:
    * Added `OLMO3_MAXTEXT_TO_HF_PARAM_MAPPING` and hooks in `param_mapping.py`.
    * Implemented identity hooks for Norms to preserve the specific scaling used in Olmo3 checkpoints.

# Tests

Tested via checkpoint conversion from Hugging Face and running `forward_pass_logit_checker.py` to verify KL divergence against the reference HF implementation (BF16 and FP32).

**1. Checkpoint Conversion:**
```bash
python3 -m MaxText.utils.ckpt_conversion.to_maxtext MaxText/configs/base.yml \
    model_name=olmo3-7b \
    hf_access_token=${HF_TOKEN} \
    base_output_directory=${BASE_OUTPUT_DIRECTORY} \
    scan_layers=True
```

**2. Logit Verification (Olmo3-7B):**

```bash
python3 -m tests.utils.forward_pass_logit_checker src/MaxText/configs/base.yml \
    model_name=olmo3-7b \
    load_parameters_path=${CHECKPOINT_PATH} \
    tokenizer_path="allenai/Olmo-3-7B-Instruct" \
    hf_model_path="allenai/Olmo-3-7B-Instruct" \
    run_hf_model=True \
    max_kl_div=0.005 \
    scan_layers=True \
    normalize_embedding_logits=False
```

Tested both Olmo3-7b and Olmo3-32 logits:
`Max KL divergence for a single token in the set: 0.000005`

https://paste.googleplex.com/4668070767493120
https://paste.googleplex.com/5281833875013632

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by dirkgr** [comment]
We talked about this in a call, but one important clarification: Olmo does not use a "mixed RoPE strategy" during pretraining. We pre-train with a context window of 8192 on the full-attention layers, and a context window of 4096 on the sliding window attention layers. No special RoPE sauce at this point. Then, after pre-training, when we do the long-context extension, we use Yarn to extend RoPE on the full attention layers, but we leave the sliding window attention layers the same.

File: tests/utils/forward_pass_logit_checker.py [hunk 1]
```diff
@@ -393,7 +397,21 @@ def main(config, test_args):  # pylint: disable=W0621
...
+    dtype_mapping = {
+        "bfloat16": torch.bfloat16,
+        "float32": torch.float32,
+        "float16": torch.float16,
+        jnp.bfloat16: torch.bfloat16,
+        jnp.float32: torch.float32,
+        jnp.float16: torch.float16,
+    }
+
+    # Default to bfloat16 if dtype is unrecognized
```
[the remark this answers is not in the document]
Comment (dirkgr) [reply tests/utils/forward_pass_logit_checker.py:410]: What is this cheerleading bot?

[the remark this answers is not in the document]
Comment (RissyRan) [reply tests/utils/forward_pass_logit_checker.py:410]: lol, Gemini CLI PR review: https://github.com/google-github-actions/run-gemini-cli

**Comment by RissyRan** [comment]
> We talked about this in a call, but one important clarification: Olmo does not use a "mixed RoPE strategy" during pretraining. We pre-train with a context window of 8192 on the full-attention layers, and a context window of 4096 on the sliding window attention layers. No special RoPE sauce at this point. Then, after pre-training, when we do the long-context extension, we use Yarn to extend RoPE on the full attention layers, but we leave the sliding window attention layers the same.

To match the logit test, it seems we need to do this in decoder layer. Gagik, do you think we could add a if/else branch to indicate pre/post training mode in this case?

File: src/maxtext/configs/types.py [hunk 2]
```diff
@@ -239,8 +239,8 @@ class ProfilerType(str, Enum):
     "gpt-oss-120b",
     "llama4-17b-16e",
     "llama4-17b-128e",
-    "olmo3_7b",
-    "olmo3_32b",
+    "olmo3-7b",
```
Comment (RissyRan) [inline src/maxtext/configs/types.py:250]: Thanks!

**Review by RissyRan — COMMENTED** [review]
LGTM, just minor comments related to that if/else for rope.

File: src/maxtext/configs/models/olmo3-7b.yml [hunk 3]
```diff
@@ -37,15 +37,18 @@ sliding_window_size: 4096
 inhomogeneous_layer_cycle_interval: 4
 
 # RoPE
-rope_type: "yarn"
+# Yarn RoPE for global attention, default RoPE for sliding window attention (set from olmo3.py)
+rope_interleave: False
```
Comment (RissyRan) [inline src/maxtext/configs/models/olmo3-7b.yml:41]: Potentially move inside of olmo3 decoder layer? based on that pre/post-training.

**Comment by dirkgr** [comment]
You needed to do that because you grabbed a model from Huggingface that had already undergone context extension. We did not use this during pre-training. Gagik knows what to do :-)

File: src/maxtext/configs/models/olmo3-7b.yml (hunk 3 above)
Comment (gagika) [reply src/maxtext/configs/models/olmo3-7b.yml:41]: created a new config for pre-training with default RoPE

File: src/MaxText/layers/attentions.py [hunk 4]
```diff
@@ -490,18 +497,28 @@ def __init__(
       self.sinks = None
 
     is_llama4_decoder_block = self.config.decoder_block == DecoderBlockType.LLAMA4
+
     if self.use_qk_norm and not is_llama4_decoder_block:
-      self.query_norm = RMSNorm(
-          num_features=self.head_dim,
+      # Check if this is Olmo3, which uses a unique "Global" QK Norm strategy.
```
[the remark this answers is not in the document]
Comment (gagika) [reply src/MaxText/layers/attentions.py:502]: good suggestion, some other models also use this convention, we need to refarctor at some point

**Comment by gagika** [comment]
> We talked about this in a call, but one important clarification: Olmo does not use a "mixed RoPE strategy" during pretraining. We pre-train with a context window of 8192 on the full-attention layers, and a context window of 4096 on the sliding window attention layers. No special RoPE sauce at this point. Then, after pre-training, when we do the long-context extension, we use Yarn to extend RoPE on the full attention layers, but we leave the sliding window attention layers the same.

@dirkgr I added a new pretraining config  olmo3-7b-pt.yml with default RoPE.

I tested logit matching against: https://huggingface.co/allenai/Olmo-3-1025-7B/tree/stage1-step1413814

```
python3 -m MaxText.utils.ckpt_conversion.to_maxtext MaxText/configs/base.yml \
    model_name=olmo3-7b-pt \
    hf_access_token=${HF_TOKEN} \
    base_output_directory=${BASE_OUTPUT_DIRECTORY} \
    scan_layers=True \
    use_multimodal=false \
    hardware=cpu \
    skip_jax_distributed_system=true \
    --revision "stage1-step1413814"
```

```
  python3 -m tests.utils.forward_pass_logit_checker src/MaxText/configs/base.yml \
    tokenizer_path=${TOKENIZER} \
    load_parameters_path=$OLMO_CHECKPOINT \
    model_name=olmo3-7b-pt \
    scan_layers=true \
    max_prefill_predict_length=16 \
    max_target_length=16 \
    use_multimodal=false \
    attention=dot_product \
    dtype=float32 \
    weight_dtype=float32 \
    --hf_model_path=$HF_LOCAL_PATH \
    --run_hf_model=True \
    --max_kl_div=0.001
```

The logits match quite well for short inputs, I will test longer text and share top 1 match rate.

cc @RissyRan

File: src/maxtext/configs/types.py (hunk 2 above)
Comment (gagika) [reply src/maxtext/configs/types.py:250]: sure

**Review by RissyRan — APPROVED** [review]
